src/textual/css/parse.py

```diff
@@ -205,9 +204,102 @@ def parse_declarations(css: str, path: str) -> Styles:
     return styles_builder.styles
 
 
-def parse(css: str, path: str) -> Iterable[RuleSet]:
+def _is_whitespace(token: Token) -> bool:
+    return token.name == "whitespace"
+
+
+def _unresolved(
+    variable_name: str, location: tuple[int, int]
+) -> UnresolvedVariableError:
+    line_idx, col_idx = location
+    return UnresolvedVariableError(
+        f"reference to undefined variable '${variable_name}' at line {line_idx + 1}, column {col_idx + 1}."
+    )
 
-    tokens = iter(tokenize(css, path))
+
+def substitute_references(tokens: Iterator[Token]) -> Iterable[Token]:
+    """Replace variable references with values by substituting variable reference
+    tokens with the tokens representing their values.
+
+    Args:
+        tokens (Iterator[Token]): Iterator of Tokens which may contain tokens
+            with the name "variable_ref".
+
+    Returns:
+        Iterable[Token]: Yields Tokens such that any variable references (tokens where
+            token.name == "variable_ref") have been replaced with the tokens representing
+            the value. In other words, an Iterable of Tokens similar to the original input,
+            but with variables resolved. Substituted tokens will have their referenced_by
+            attribute populated with information about where the tokens are being substituted to.
+    """
+    variables: dict[str, list[Token]] = defaultdict(list)
+    while tokens:
+        token = next(tokens, None)
+        if token is None:
+            break
+        if token.name == "variable_name":
+            variable_name = token.value[1:-1]  # Trim the $ and the :, i.e. "$x:" -> "x"
+            yield token
+
+            # Store the tokens for any variable definitions, and substitute
+            # any variable references we encounter with them.
+            leading_whitespace = True
+            while True:
+                token = next(tokens, None)
+                if not token:
+                    break
+                elif leading_whitespace and token.name == "whitespace":
+                    yield token
+                    leading_whitespace = False
+                elif token.name == "variable_value_end":
+                    yield token
+                    break
+                # For variables referring to other variables
+                elif token.name == "variable_ref":
+                    ref_name = token.value[1:]
+                    if ref_name in variables:
+                        variable_tokens = variables[variable_name]
+                        reference_tokens = variables[ref_name]
+                        variable_tokens.extend(reference_tokens)
+                        ref_location = token.location
+                        ref_length = cell_len(token.value)
```

Comment:
Is this definitely the cell length you need here? I would have thought it would be the number of characters that is relevant.

Reply:
This was making the assumption that variables of "string" type could have non-ASCII values, which admittedly I'm not sure about. If the value is out-with ASCII, don't we want to know how many cells wide it is so that we can highlight all of those cells? If we take the `len`, we may end up highlighting more cells than required.